### PR TITLE
Nccl2 dist API

### PR DESCRIPTION
**Pull request by typhoonzero**

### CODE REVIEW

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 1]
```diff
@@ -163,6 +168,18 @@ class DistributeTranspiler(object):
                                                                 pserver_program)
            elif role == "TRAINER":
                 trainer_program = t.get_trainer_program()
+           
+           # for nccl2 mode
+           config = fluid.DistributeTranspilerConfig()
+           config.mode = "nccl2"
```
Comment (panyx0718) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:184]: ring might be a better name

**Review by panyx0718 — DISMISSED** [review]

**Review by reyoung — DISMISSED** [review]

File: python/paddle/fluid/transpiler/distribute_transpiler.py [hunk 2]
```diff
@@ -163,6 +168,18 @@ class DistributeTranspiler(object):
...
+           t = fluid.DistributeTranspiler(config=config)
+           t.transpile(trainer_id, workers=workers, current_endpoint=curr_ep)
```
Comment (kolinwei) [inline python/paddle/fluid/transpiler/distribute_transpiler.py:186]: workers？it should be trainers?

Comment (typhoonzero) [reply python/paddle/fluid/transpiler/distribute_transpiler.py:186]: Will update in next PR since this PR need API approvals

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by panyx0718 — APPROVED** [review]

**Review by qingqing01 — APPROVED** [review]

**Pull request merged**